### PR TITLE
add functional tests for reset activity API

### PR DESCRIPTION
**Pull request by ychebotarev**

### CODE REVIEW

File: tests/activity_api_reset_test.go [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+	}
+}
+
+func (s *ActivityApiResetClientTestSuite) TestActivityResetApi_Acceptance() {
+	// activity reset is called after multiple attempts,
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+
+	var activityReseted atomic.Bool
+	activityCompleteSignal := make(chan struct{})
```
Comment (stephanos) [inline tests/activity_api_reset_test.go:103]: nit: maybe use `activityCompleteCh`? (since "Signal" is a bit overloaded within Temporal)

_(same below)_

File: tests/activity_api_reset_test.go [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+	var activityCompleted atomic.Int32
+
+	activityFunction := func() (string, error) {
+		if activityReseted.Load() == false {
+			activityErr := errors.New("bad-luck-please-retry")
+			activityCompleted.Add(1)
+			return "", activityErr
+		}
+
+		activityCompleted.Add(1)
```
Comment (stephanos) [inline tests/activity_api_reset_test.go:113]: nit: you could change these two calls to a single `defer activityCompleted.Add(1)` at the top of the activity function

_(same below)_

File: tests/activity_api_reset_test.go [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+		return "done!", err
+	}
+}
+
+func (s *ActivityApiResetClientTestSuite) TestActivityResetApi_Acceptance() {
+	// activity reset is called after multiple attempts,
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+
+	var activityReseted atomic.Bool
```
Comment (stephanos) [inline tests/activity_api_reset_test.go:102]: nit: maybe `activityWasReset`? (just for grammar)

File: tests/activity_api_reset_test.go [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+	activityCompleteSignal := make(chan struct{})
+	var activityCompleted atomic.Int32
```
Comment (stephanos) [inline tests/activity_api_reset_test.go:104]: nit: maybe `activityCompletionCount`? (at first I was confused why it wasn't a boolean)

_(same below)_

File: tests/activity_api_reset_test.go [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+		assert.Equal(t, int32(1), description.PendingActivities[0].Attempt)
+	}, 5*time.Second, 100*time.Millisecond)
+
+	// let activity finish
+	activityCompleteSignal <- struct{}{}
+
+	// wait for workflow to complete
+	var out string
+	err = workflowRun.Get(ctx, &out)
+
```
Comment (stephanos) [inline tests/activity_api_reset_test.go:168]: nit: no need for newline (other error checks are right below the error)

_(same below)_

File: tests/activity_api_reset_test.go [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+	// wait for activity to start
+	s.EventuallyWithT(func(t *assert.CollectT) {
+		description, err := s.SdkClient().DescribeWorkflowExecution(ctx, workflowRun.GetID(), workflowRun.GetRunID())
+		assert.NoError(t, err)
+		assert.Equal(t, 1, len(description.PendingActivities))
+		assert.Equal(t, enumspb.PENDING_ACTIVITY_STATE_STARTED, description.PendingActivities[0].State)
+	}, 5*time.Second, 200*time.Millisecond)
+
+	activityReseted.Store(true)
```
Comment (stephanos) [inline tests/activity_api_reset_test.go:214]: the previous test had this line _after_ the `ResetActivityByIdRequest` call; does the order matter?

File: tests/activity_api_reset_test.go [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+	activityCompleteSignal <- struct{}{}
+
+	// wait for workflow to complete
+	var out string
+	err = workflowRun.Get(ctx, &out)
+
+	s.NoError(err)
+}
+
+func (s *ActivityApiResetClientTestSuite) TestActivityResetApi_WithNoWait() {
```
Comment (stephanos) [inline tests/activity_api_reset_test.go:172]: nit: maybe call it `TestActivityResetApi_WhileRunning`? since all the other tests also use `NoWait=true`, it's not unique to this test.

File: tests/activity_api_reset_test.go [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+			DisableEagerExecution:  true,
+			StartToCloseTimeout:    s.startToCloseTimeout,
+			ScheduleToCloseTimeout: s.scheduleToCloseTimeout,
+			RetryPolicy:            s.activityRetryPolicy,
+		}), activityFunction).Get(ctx, &ret)
+		return "done!", err
+	}
+}
+
+func (s *ActivityApiResetClientTestSuite) TestActivityResetApi_Acceptance() {
```
Comment (stephanos) [inline tests/activity_api_reset_test.go:97]: nit: maybe call it `TestActivityResetApi_AfterRetries` or sth?

**Review by stephanos — APPROVED** [review]

File: tests/activity_api_reset_test.go (hunk 1 above)
Comment (ychebotarev) [reply tests/activity_api_reset_test.go:103]: renamed

File: tests/activity_api_reset_test.go (hunk 2 above)
Comment (ychebotarev) [reply tests/activity_api_reset_test.go:113]: true

File: tests/activity_api_reset_test.go (hunk 3 above)
Comment (ychebotarev) [reply tests/activity_api_reset_test.go:102]: yeah, IDE even underline "reseted". will change

File: tests/activity_api_reset_test.go (hunk 4 above)
Comment (ychebotarev) [reply tests/activity_api_reset_test.go:104]: done

File: tests/activity_api_reset_test.go (hunk 5 above)
Comment (ychebotarev) [reply tests/activity_api_reset_test.go]: removed

File: tests/activity_api_reset_test.go (hunk 7 above)
Comment (ychebotarev) [reply tests/activity_api_reset_test.go]: renamed

File: tests/activity_api_reset_test.go (hunk 8 above)
Comment (ychebotarev) [reply tests/activity_api_reset_test.go]: renamed

File: tests/activity_api_reset_test.go (hunk 6 above)
Comment (ychebotarev) [reply tests/activity_api_reset_test.go]: yes, in a subtle way

File: tests/activity_api_reset_test.go [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+	}
+
+	workflowRun, err := s.SdkClient().ExecuteWorkflow(ctx, workflowOptions, workflowFn)
+	s.NoError(err)
+
+	// wait for activity to start/fail few times
+	s.EventuallyWithT(func(t *assert.CollectT) {
+		description, err := s.SdkClient().DescribeWorkflowExecution(ctx, workflowRun.GetID(), workflowRun.GetRunID())
+		assert.NoError(t, err)
+		assert.Equal(t, 1, len(description.PendingActivities))
```
Comment (alexshtin) [inline tests/activity_api_reset_test.go:136]: What I recently figured out, is that `assert.xxx(t...` doesn't actually fail the test but continue execution (which is kinda obvious if you think about it). So if `DescribeWorkflowExecution` returns `err` and `nil` description, this will panic with NPE on line 136.

File: tests/activity_api_reset_test.go [hunk 10]
```diff
@@ -0,0 +1,325 @@
...
+)
+
+type ActivityApiResetClientTestSuite struct {
+	testcore.ClientFunctionalSuite
+	tv                     *testvars.TestVars
+	initialRetryInterval   time.Duration
+	scheduleToCloseTimeout time.Duration
+	startToCloseTimeout    time.Duration
+
+	activityRetryPolicy *temporal.RetryPolicy
```
Comment (alexshtin) [inline tests/activity_api_reset_test.go:55]: I don't like all of this as fields of suite struct. The need to be set in every test and passed explicitly to `makeWorkflowFunc`.

File: tests/activity_api_reset_test.go [hunk 11]
```diff
@@ -0,0 +1,325 @@
...
+	s.OverrideDynamicConfig(dynamicconfig.ActivityAPIsEnabled, true)
+	s.tv = testvars.New(s.T()).WithTaskQueue(s.TaskQueue()).WithNamespaceName(namespace.Name(s.Namespace()))
+}
+
+func (s *ActivityApiResetClientTestSuite) SetupTest() {
+	s.ClientFunctionalSuite.SetupTest()
+
+	s.initialRetryInterval = 1 * time.Second
+	s.scheduleToCloseTimeout = 30 * time.Minute
+	s.startToCloseTimeout = 15 * time.Minute
```
Comment (alexshtin) [inline tests/activity_api_reset_test.go:69]: This means "infinity" right? There is `tv.InfinityTimeout()` for it.

File: tests/activity_api_reset_test.go [hunk 12]
```diff
@@ -0,0 +1,325 @@
...
+		s.WaitForChannel(ctx, activityCompleteCh)
+		return "done!", nil
+	}
+
+	workflowFn := s.makeWorkflowFunc(activityFunction)
+
+	s.Worker().RegisterWorkflow(workflowFn)
+	s.Worker().RegisterActivity(activityFunction)
+
+	wfId := testcore.RandomizeStr("wfid-" + s.T().Name())
```
Comment (alexshtin) [inline tests/activity_api_reset_test.go:123]: If we decided to use `tv` (and your seems to have it here too), it should be used for all things like this. And also for all other pseudo-random strings that you have in these tests.

File: tests/activity_api_reset_test.go [hunk 13]
```diff
@@ -0,0 +1,325 @@
...
+	}
+}
+
+func TestActivityApiResetClientTestSuite(t *testing.T) {
+	s := new(ActivityApiResetClientTestSuite)
+	suite.Run(t, s)
+}
+
+func (s *ActivityApiResetClientTestSuite) makeWorkflowFunc(activityFunction ActivityFunctions) WorkflowFunction {
+	return func(ctx workflow.Context) (string, error) {
```
Comment (alexshtin) [inline tests/activity_api_reset_test.go:83]: You don't have to return `string` if you don't really check it. Just `error` would be good enough.

File: tests/activity_api_reset_test.go (hunk 10 above)
Comment (bergundy) [reply tests/activity_api_reset_test.go:55]: +1 this makes it hard to follow the test code and figure out which values are applied. Better apply at the call site.

File: tests/activity_api_reset_test.go [hunk 14]
```diff
@@ -0,0 +1,325 @@
...
+	scheduleToCloseTimeout time.Duration
+	startToCloseTimeout    time.Duration
+
+	activityRetryPolicy *temporal.RetryPolicy
+}
+
+func (s *ActivityApiResetClientTestSuite) SetupSuite() {
+	s.ClientFunctionalSuite.SetupSuite()
+	s.OverrideDynamicConfig(dynamicconfig.ActivityAPIsEnabled, true)
+	s.tv = testvars.New(s.T()).WithTaskQueue(s.TaskQueue()).WithNamespaceName(namespace.Name(s.Namespace()))
```
Comment (bergundy) [inline tests/activity_api_reset_test.go:61]: @alexshtin we shouldn't have to do this. Maybe put a `TV()` method on the base test suite struct?

File: tests/activity_api_reset_test.go (hunk 12 above)
Comment (bergundy) [reply tests/activity_api_reset_test.go:123]: +1

File: tests/activity_api_reset_test.go (hunk 9 above)
Comment (bergundy) [reply tests/activity_api_reset_test.go:136]: Very true! Good catch.

You can work around this with:

```go
		assert.Len(t, description.GetPendingActivities(), 1)
```

File: tests/activity_api_reset_test.go [hunk 15]
```diff
@@ -0,0 +1,325 @@
...
+	s.NotNil(resp)
+
+	activityWasReset.Store(true)
+
+	// wait for activity to be running
+	s.EventuallyWithT(func(t *assert.CollectT) {
+		description, err := s.SdkClient().DescribeWorkflowExecution(ctx, workflowRun.GetID(), workflowRun.GetRunID())
+		assert.NoError(t, err)
+		assert.Equal(t, 1, len(description.PendingActivities))
+		assert.Equal(t, enumspb.PENDING_ACTIVITY_STATE_STARTED, description.PendingActivities[0].State)
```
Comment (bergundy) [inline tests/activity_api_reset_test.go:157]: Same as above.

File: tests/activity_api_reset_test.go [hunk 16]
```diff
@@ -0,0 +1,325 @@
...
+
+	workflowRun, err := s.SdkClient().ExecuteWorkflow(ctx, workflowOptions, workflowFn)
+	s.NoError(err)
+
+	// wait for activity to start/fail few times
+	s.EventuallyWithT(func(t *assert.CollectT) {
+		description, err := s.SdkClient().DescribeWorkflowExecution(ctx, workflowRun.GetID(), workflowRun.GetRunID())
+		assert.NoError(t, err)
+		assert.Equal(t, 1, len(description.PendingActivities))
+		assert.True(t, activityCompleteCount.Load() > 1)
```
Comment (bergundy) [inline tests/activity_api_reset_test.go:137]: Use assert.GreatOrEqual to get a better error message when this fails.

File: tests/activity_api_reset_test.go (hunk 15 above)
Comment (bergundy) [inline tests/activity_api_reset_test.go:157]: You can get rid of the Eventually block and just check that the attempt was reset. That should be immediate.

The state transitioning to STARTED can be eventual and has no value for this test. You already check that the activity completes so it must transition to STARTED to do that.

File: tests/activity_api_reset_test.go [hunk 17]
```diff
@@ -0,0 +1,325 @@
...
+
+	// wait for activity to start
+	s.EventuallyWithT(func(t *assert.CollectT) {
+		description, err := s.SdkClient().DescribeWorkflowExecution(ctx, workflowRun.GetID(), workflowRun.GetRunID())
+		assert.NoError(t, err)
+		assert.Equal(t, 1, len(description.PendingActivities))
+		assert.Equal(t, enumspb.PENDING_ACTIVITY_STATE_STARTED, description.PendingActivities[0].State)
+	}, 5*time.Second, 200*time.Millisecond)
+
+	activityWasReset.Store(true)
```
Comment (bergundy) [inline tests/activity_api_reset_test.go:213]: This should only be set after the activity was reset, not before the call.

File: tests/activity_api_reset_test.go [hunk 18]
```diff
@@ -0,0 +1,325 @@
...
+	workflowRun, err := s.SdkClient().ExecuteWorkflow(ctx, workflowOptions, workflowFn)
+	s.NoError(err)
+
+	// wait for activity to start, fail and wait for retry
+	s.EventuallyWithT(func(t *assert.CollectT) {
+		description, err := s.SdkClient().DescribeWorkflowExecution(ctx, workflowRun.GetID(), workflowRun.GetRunID())
+		assert.NoError(t, err)
+		assert.Equal(t, 1, len(description.PendingActivities))
+		assert.Equal(t, enumspb.PENDING_ACTIVITY_STATE_SCHEDULED, description.PendingActivities[0].State)
+		assert.Equal(t, int32(1), activityCompleteCount.Load())
```
Comment (bergundy) [inline tests/activity_api_reset_test.go:294]: Same as my review yesterday, this could be more than 1 in a busy CI.

File: tests/activity_api_reset_test.go [hunk 19]
```diff
@@ -0,0 +1,325 @@
...
+		WorkflowId: workflowRun.GetID(),
+		ActivityId: "activity-id",
+		NoWait:     true,
+	}
+	resp, err := s.FrontendClient().ResetActivityById(ctx, resetRequest)
+	s.NoError(err)
+	s.NotNil(resp)
+
+	// wait for activity to start. Wait time is shorter than original retry interval
+	s.EventuallyWithT(func(t *assert.CollectT) {
```
Comment (bergundy) [inline tests/activity_api_reset_test.go:308]: Same as above, just check that the attempt was reset to 1 without an eventually block.

File: tests/activity_api_reset_test.go (hunk 11 above)
Comment (ychebotarev) [reply tests/activity_api_reset_test.go:69]: tv.InfinityTimeout() has different type

File: tests/activity_api_reset_test.go (hunk 15 above)
Comment (ychebotarev) [reply tests/activity_api_reset_test.go:157]: no I can't get rid of this Eventually block. 
To check that the number of attempts was reset I need to call DescribeWorkflowExecution.
And this may not happen instantly after ResetActivityById call

File: tests/activity_api_reset_test.go (hunk 17 above)
Comment (ychebotarev) [reply tests/activity_api_reset_test.go:213]: no, it should be set here. but naming may be missleading, I will rename it
